Added new filter to HistoricTaskInstanceQuery

parth-lunagariya opened the pull request:
Added new filter taskAssigned(filtering task where assignee is not null) and taskUnassigned(filtering task where assignee is null) for HistoricTaskInstanceQuery

filiphr commented:
Thanks a lot for the PR @parth-lunagariya. I think that this is a useful addition.

However, I think that it makes more sense to add this to the `TaskInfoQuery` so it is available for runtime tasks as well. Additionally, would you mind adding some tests for this as well. You can add them to [`TaskInfoQueryTest`](https://github.com/flowable/flowable-engine/blob/master/modules/flowable-engine/src/test/java/org/flowable/engine/test/api/task/TaskInfoQueryTest.java) and [`HistoricTaskInstanceTest`](https://github.com/flowable/flowable-engine/blob/master/modules/flowable-engine/src/test/java/org/flowable/engine/test/history/HistoricTaskInstanceTest.java).

Finally, if you want to make this available through the REST API you can have a look the the usages of [`TaskBaseResource#getTasksFromQueryRequest`](https://github.com/flowable/flowable-engine/blob/9a102f654d8280e93b871d3e21400645707741fc/modules/flowable-rest/src/main/java/org/flowable/rest/service/api/runtime/task/TaskBaseResource.java#L127) and [`HistoricTaskInstanceBaseResource#getQueryResponse`](https://github.com/flowable/flowable-engine/blob/9a102f654d8280e93b871d3e21400645707741fc/modules/flowable-rest/src/main/java/org/flowable/rest/service/api/history/HistoricTaskInstanceBaseResource.java#L74) for the BPMN engine and [`TaskBaseResource#getTasksFromQueryRequest`](https://github.com/flowable/flowable-engine/blob/9a102f654d8280e93b871d3e21400645707741fc/modules/flowable-cmmn-rest/src/main/java/org/flowable/cmmn/rest/service/api/runtime/task/TaskBaseResource.java#L127) and [`HistoricTaskInstanceBaseResource#getQueryResponse`](https://github.com/flowable/flowable-engine/blob/9a102f654d8280e93b871d3e21400645707741fc/modules/flowable-cmmn-rest/src/main/java/org/flowable/cmmn/rest/service/api/history/task/HistoricTaskInstanceBaseResource.java#L74) for the CMMN engine.

parth-lunagariya commented:
Hi @filiphr ,

I haven't added those filter in TaskInfoQuery because runtime tasks already have similar filter "taskUnassigned" in TaskQuery.java file which is i think specific for runtime task and now I want that for historic task also for my project. I haven't added in REST API as we are not using REST API and i might break something in REST API.

I will add required test cases for HistoricTaskInstanceTest.

filiphr commented:
I forgot about the `TaskQuery#taskUnassigned` thanks for reminding me. In that case I would suggest moving method in the `TaskInfoQuery` as the implementation logic is identical. The method should instead return `T` (the generic argument of the `TaskInfoQuery`).

Let's also move `taskAssigned` to the `TaskInfoQuery`, because people will surely want to use that as well 😄.

No problem for the REST API, we'll have a look at it once you are done.

parth-lunagariya commented:
ok @filiphr, I will move both methods to TaskInfoQuery as you said and I will also change necessary test cases.

parth-lunagariya commented:
I have moved both methods to TaskInfoQuery. I also changed required test cases. I haven't change TaskInfoQueryTest as it looks like more generic test case. so I added test case in test file of implementation of that query(TaskQueryTest and HistoricTaskInstanceTest).

@filiphr  please review this code changes.

parth-lunagariya commented:
> I have moved both methods to TaskInfoQuery. I also changed required test cases. I haven't change TaskInfoQueryTest as it looks like more generic test case. so I added test case in test file of implementation of that query(TaskQueryTest and HistoricTaskInstanceTest).
> 
> @filiphr please review this code changes.

Hi @filiphr , I did required code changes. Let me know any more changes required.

filiphr commented on modules/flowable-engine/src/test/java/org/flowable/engine/test/api/task/TaskQueryTest.java:
```diff
@@ -1268,6 +1268,20 @@ public void testQueryByUnassignedOr() {
+    @Test
+    public void testQueryByAssignedOr() {
+        TaskQuery query = taskService.createTaskQuery().or().taskId("invalid").taskAssigned();
+        assertThat(query.count()).isEqualTo(0);
```
It seems like the `or` is not working correctly. This should have a count of `1` since there is 1 assigned task.

parth-lunagariya replied:
Fixed above issue.

filiphr commented on modules/flowable-task-service/src/main/resources/org/flowable/task/service/db/mapping/entity/Task.xml:
```diff
@@ -1123,6 +1126,9 @@
               or RES.ASSIGNEE_ IS NULL
             </if>
+            <if test="orQueryObject.withAssignee">
+              and RES.ASSIGNEE_ IS NOT NULL
```
This should be an `or` and not an `and`.

parth-lunagariya replied:
ohhh. i will fix this issue and let you know @filiphr.

https://github.com/flowable/flowable-engine/pull/2756 this PR is also created by me. can you please look at that?

parth-lunagariya replied:
Fixed above issue.

filiphr commented:
@parth-lunagariya, I was on holidays, that's why the slow response.

Anyways, I've left 2 comments, can you please do those fixes and then we can merge this.

parth-lunagariya commented:
@filiphr , I made requested code changes. Please review.

filiphr commented:
Thanks for the changes @parth-lunagariya. I've merged this manually in https://github.com/flowable/flowable-engine/commit/0847d9dd8f5739bc2503a7bd80598597411a1190
